smtp.go

```diff
@@ -2,209 +2,391 @@ package maddy
 
 import (
 	"bytes"
+	"crypto/tls"
 	"errors"
 	"fmt"
 	"io"
+	"net"
+	"net/mail"
 	"os"
 	"strings"
-	"time"
+	"sync"
 
-	"github.com/emersion/go-msgauth"
-	pgppubkey "github.com/emersion/go-pgp-pubkey"
-	pgppubkeylocal "github.com/emersion/go-pgp-pubkey/local"
-	pgplocal "github.com/emersion/go-pgpmail/local"
-	smtppgp "github.com/emersion/go-pgpmail/smtp"
 	"github.com/emersion/go-smtp"
-	smtpproxy "github.com/emersion/go-smtp-proxy"
-	"github.com/emersion/go-smtp/backendutil"
-	"github.com/mholt/caddy/caddyfile"
-	"golang.org/x/crypto/openpgp"
+	"github.com/emersion/maddy/config"
+	"github.com/emersion/maddy/log"
+	"github.com/emersion/maddy/module"
 )
 
-func newSMTPServer(tokens map[string][]caddyfile.Token, hostname string, lmtp bool) (server, error) {
-	be, err := newUpstream(tokens)
-	if err != nil {
-		return nil, err
+type SMTPUser struct {
+	endp      *SMTPEndpoint
+	anonymous bool
+	username  string
+	state     *smtp.ConnectionState
+}
+
+func (u SMTPUser) Send(from string, to []string, r io.Reader) error {
```

Comment:
I'd prefer these to take a pointer. It's too easy to be surprised by `u` not being effectively mutated when doing `u.thing = stuff` in a method.